Windows: Fix module "file.check_file_meta" incorrect diff

skizunov opened the pull request:
### What does this PR do?

On Windows, the module function `file.check_file_meta` would always
return a diff even when the files are identical. For instance, the files
are identical but this is returned:

```
"pchanges": {
  "diff": "--- \n+++ \n@@ -1 +1 @@\n
    -{\"Skyline.ExchangeName\":\"niskyline\",\"Skyline.Host\":\"127.0.0.1\",
      \"Skyline.Password\":\"Y5{9()*Dy:x(d&m:FKnmPFb+b5!!!Ya)\",
      \"Skyline.User\":\"niskyline\"}
    +{\"Skyline.ExchangeName\":\"niskyline\",\"Skyline.Host\":\"127.0.0.1\",
      \"Skyline.Password\":\"Y5{9()*Dy:x(d&m:FKnmPFb+b5!!!Ya)\",
      \"Skyline.User\":\"niskyline\"}\n"
}
```

This type of issue was already fixed within similar logic in
`file.manage_file` by using:

```
if salt.utils.is_windows():
    contents = os.linesep.join(contents.splitlines())
```

Add this exact same logic to `file.check_file_meta`.
### Tests written?

No
Signed-off-by: Sergey Kizunov sergey.kizunov@ni.com

skizunov committed:
Windows: Fix module "file.check_file_meta" incorrect diff

On Windows, the module function `file.check_file_meta` would always
return a diff even when the files are identical. For instance, the files
are identical but this is returned:

```
"pchanges": {
  "diff": "--- \n+++ \n@@ -1 +1 @@\n
    -{\"Skyline.ExchangeName\":\"niskyline\",\"Skyline.Host\":\"127.0.0.1\",
      \"Skyline.Password\":\"Y5{9()*Dy:x(d&m:FKnmPFb+b5!!!Ya)\",
      \"Skyline.User\":\"niskyline\"}
    +{\"Skyline.ExchangeName\":\"niskyline\",\"Skyline.Host\":\"127.0.0.1\",
      \"Skyline.Password\":\"Y5{9()*Dy:x(d&m:FKnmPFb+b5!!!Ya)\",
      \"Skyline.User\":\"niskyline\"}\n"
}
```

This type of issue was already fixed within similar logic in
`file.manage_file` by using:

```
if salt.utils.is_windows():
    contents = os.linesep.join(contents.splitlines())
```

Add this exact same logic to `file.check_file_meta`.

Signed-off-by: Sergey Kizunov <sergey.kizunov@ni.com>

damon-atkins commented:
seems not to be the fix. Need to work out why they are coming across as different. e.g open options.

skizunov commented:
@damon-atkins: Do you have an example where it still fails? My example was a one-liner and it fixed that. The main reason why it fixed the one-liner is probably because:

```
contents = os.linesep.join(contents.splitlines())
```

Will remove the trailing newline at the very end. So adding this is necessary because when it writes the original to file, it does the same thing. Let me try this on a multi-line file and get back to you to see if we also need to change the file write mode from "wb" to "w" for the temp file.

skizunov committed:
Change file.check_file_meta write mode to "w"

In order to be consistent with how `file.manage_file` writes a file,
we make `file.check_file_meta` also write the file as "w" instead of "wb".
This is necessary so that diff functionality correctly identifies
identical files when the source file is multi-line.

Signed-off-by: Sergey Kizunov <sergey.kizunov@ni.com>

skizunov commented:
@damon-atkins : You were right. Multi-line files were still broken. To fix it, both the original change is necessary as well as changing the write mode from "wb" to "w". I added a commit to fix the multi-line issue.

damon-atkins commented:
looks like you just need to read both as binary. I would have thought that the line feed on the salt master would be the same on the client.

e.g.  if you want windows line feeds, the file on the salt master should have windows line feeds.  Salt should not be doing any line feed translation.  As that is altering the file and can have unexpected results

skizunov commented:
@damon-atkins : Reading both as binary doesn't work. It just complicates things because everything needs to be converted from `bytes` back to `str` (in Python 3 note that when you read as binary, it will return a `bytes` object but when you read as non-binary it will return a `str` object) before the diff library will work (the diff library requires `str` objects and will fail if you pass in `bytes` objects in Python 3). But even if you read both as binary, it will diff same files different because it is written as non-binary in one case (the original file in `file.manage_file`) and was originally written as binary in the other case (the temp file in `file.check_file_meta`). So the file contents on disk are different (on Windows). The solution that was submitted (and already merged) is probably the best solution simply because it matches exactly what `file.manage_file` does.

I did try a bunch of variants beforehand including the read as binary suggestion that you gave (which I discarded as a viable solution due to the reasons above).

damon-atkins commented:
salt/modules/file.py This soft of indicates this is not available on windows, yet their is a fair bit of `salt.utils.is_windows()` spread through out the code.  I must be missing something.
```
def __virtual__():
    '''
    Only work on POSIX-like systems
    '''
    # win_file takes care of windows
    if salt.utils.is_windows():
        return (False, 'The file execution module cannot be loaded: only available on non-Windows systems - use win_file instead.')
    return True
```

skizunov commented:
@damon-atkins : This is a trick that Salt uses on a number of different modules, including this one. If you look at `win_file.py`:

```
from salt.modules.file import (check_hash,  # pylint: disable=W0611
        directory_exists, get_managed, mkdir, makedirs_, makedirs_perms,
        check_managed, check_managed_changes, check_perms, source_list,
        touch, append, contains, contains_regex,
        contains_glob, find, psed, get_sum, _get_bkroot, _mkstemp_copy,
        get_hash, manage_file, file_exists, get_diff, line, list_backups,
        __clean_tmp, check_file_meta, _binary_replace, restore_backup,
        access, copy, readdir, rmdir, truncate, replace, delete_backup,
        search, _get_flags, extract_hash, _error, _sed_esc, _psed,
        RE_FLAG_TABLE, blockreplace, prepend, seek_read, seek_write, rename,
        lstat, path_exists_glob, write, pardir, join, HASHES, comment,
        uncomment, _add_flags, comment_line, apply_template_on_contents)
```

So even though `win_file.py` is used, `win_file.py` still loads a lot of code from `file.py`.

damon-atkins commented:
It would be nicer if file_share.py existed and both  posix_file.py loads an win_file.py imports it..  I notice the trick it just before I read your post.   

I will raise this as a separate issue as its bigger than this PR.
> There are multiple competing Line Ending Formats:
> Lines ending with LF+CR
> Lines ending with LF
> Or even CR and CR+LF
> It's a bad assumptions on Python part that only one type will be on the platform.
> Application A on windows may use LF+CR and application B may use LF.   Or on Unix you may use salt to generate a file on Unix, which is provide to windows clients using Samba.
> Text handling needs to hand all cases on a platform.   


Thanks @skizunov
